android/src/main/java/com/swmansion/reanimated/nodes/ProceduralNode.java

```diff
@@ -0,0 +1,144 @@
+      }
+    }
+
+    @Override
+    protected Object evaluate(EvalContext oldEvalContext) {
+      if (mOldContext == null) {
+        ProceduralNode proceduralNode =  mNodesManager.findNodeById(mProceduralNode, ProceduralNode.class);
+        mOldContext = oldEvalContext;
+        for (int i = 0; i < mArgumentsInputs.length; i++) {
+          int argumentID = proceduralNode.mProceduralArguments[i];
+          ArgumentNode arg = mNodesManager.findNodeById(argumentID, ArgumentNode.class);
+          Node inputNode = mNodesManager.findNodeById(mArgumentsInputs[i], Node.class);
+          arg.matchContextWithNode(
+                  mEvalContext,
+                  inputNode
+          );
+          arg.matchNodeWithOldContext(
+                  inputNode,
+                  oldEvalContext
+          );
+        }
+      } else if (mOldContext != oldEvalContext) {
+        throw new IllegalArgumentException("Tried to evaluate perform node in more than one contexts");
+      }
+
+      ProceduralNode proceduralNode = mNodesManager.findNodeById(mProceduralNode, ProceduralNode.class);
+      return proceduralNode.value(mEvalContext);
+    }
+  }
+
+  static public class ArgumentNode extends ValueNode {
+    private final SparseArray<Node> mValuesByContext = new SparseArray<>();
+    private final SparseArray<EvalContext> mContextsByValue = new SparseArray<>();
+    private final SparseArray<EvalContext> mOldContextByValue = new SparseArray<>();
+
+    public ArgumentNode(int nodeID, ReadableMap config, NodesManager nodesManager) {
+      super(nodeID, config, nodesManager);
+    }
+
+    public void matchContextWithNode(EvalContext context, Node node) {
+      mValuesByContext.put(context.contextID, node);
+      mContextsByValue.put(node.mNodeID, context);
+    }
+
+    public void dropContext(EvalContext evalContext) {
+      Node relatedNode = mValuesByContext.get(evalContext.contextID);
+      mContextsByValue.remove(relatedNode.mNodeID);
+      mValuesByContext.remove(evalContext.contextID);
+      mOldContextByValue.remove(relatedNode.mNodeID);
+    }
+
+    public void matchNodeWithOldContext(Node node, EvalContext evalContext) {
+      mOldContextByValue.put(node.mNodeID, evalContext);
+    }
+
+    @Override
+    public EvalContext switchContextWhileUpdatingIfNeeded(EvalContext evalContext, Node lastVisited) {
+      if (lastVisited == null) {
+        return evalContext;
+      }
+      return mContextsByValue.get(lastVisited.mNodeID);
+    }
+
+    @Override
+    public void setValue(Object value, EvalContext context) {
+      ((ValueNode)mValuesByContext.get(context.contextID)).setValue(value, mNodesManager.globalEvalContext);
+    }
+
+    @Override
+    protected Double evaluate(EvalContext evalContext) {
+      if (evalContext == mNodesManager.globalEvalContext) {
+        throw new IllegalArgumentException("Tried to evaluate argumentNode in global context");
+      }
+      Node value = mValuesByContext.get(evalContext.contextID);
```

Comment:
Maybe it'd be better to create some simple struct for storing node with accompanying context instead of having two hash maps like that